awx/main/scheduler/dag_workflow.py

```diff
@@ -215,10 +214,20 @@ def mark_dnr_nodes(self):
             obj = node['node_object']
 
             if obj.do_not_run is False and not obj.job and node not in root_nodes:
-                parent_nodes = [p['node_object'] for p in self.get_dependents(obj)]
+                parent_nodes = [p['node_object'] for p in self.get_parents(obj)]
                 if self._are_all_nodes_dnr_decided(parent_nodes):
                     if self._should_mark_node_dnr(node, parent_nodes):
                         obj.do_not_run = True
                         nodes_marked_do_not_run.append(node)
+            if not obj.do_not_run and obj.all_parents_must_converge:
+                if self._are_relevant_parents_finished(node):
```

Comment:
@chrismeyersfsu when we discussed this, timing came up. Imagine a node that is marked as `all_parents_must_converge` with 2 parents. One parent has not started, and another parent has been marked as DNR. With this `if` here, the workflow will wait to process the node until the 2nd parent node has a terminal status (success/failed/DNR). But that information isn't needed to know what will happen. Any parent becoming DNR will mean that the node becomes DNR. So what I'm getting to - is it safe to propagate that status downstream to the rest of the branch at that point? I'm suggesting that instead of waiting for all parents to finish, the subsequent loop over parent nodes could handle the case of unfinished nodes, so that it could mark DNR nodes earlier.

Reply:
yes